File: 2019/01/day1.py
import sys
from typing import List

def cost(mass: int) -> int:
    return mass // 3 - 2

def cost_rec(mass: int) -> int:
    c = cost(mass)
    return 0 if c <= 0 else c + cost_rec(c)

lines = [int(l) for l in sys.stdin.readlines()]
print(f'Part 1: {sum(map(cost, lines))}')
print(f'Part 2: {sum(map(cost_rec, lines))}')

File: 2019/01/day1.rs
use std::io::{self, BufRead};
use std::ops::Div;

fn main() {
    let lines: Vec<u32> = io::stdin()
        .lock()
        .lines()
        .map(|l| l.unwrap().parse().unwrap())
        .collect();
    println!("Part 1: {}", part1(&lines));
    println!("Part 2: {}", part2(&lines));
}

fn part1(numbers: &Vec<u32>) -> u32 {
    numbers.iter().map(|&m| cost(m)).sum()
}

fn part2(numbers: &Vec<u32>) -> u32 {
    numbers.iter().map(|&m| cost_rec(m, 0)).sum()
}

fn cost(mass: u32) -> u32 {
    mass.div(3).saturating_sub(2)
}

fn cost_rec(mass: u32, acc: u32) -> u32 {
    match cost(mass) {
        0 => acc,
        c => cost_rec(c, acc + c)
    }
}
